### PR TITLE
board: add support for the STM nucleo-f334 board

### PR DESCRIPTION
**Pull request by haukepetersen**

### CODE REVIEW

**Comment by kaspar030** [comment]
The story is: Hauke sitting at POINT kickoff meeting. Someone shows him the nucleo board, asks for RIOT support. Hauke says "no, but I bet I can make a port in 30min." After 27 minutes he had hello-world running on the board.

File: boards/nucleo-f334/Makefile.include [hunk 1]
```diff
@@ -0,0 +1,56 @@
+## the cpu to build for
+export CPU = stm32f3
+export CPU_MODEL = stm32f334r8
+
+#define the default port depending on the host OS
+OS := $(shell uname)
+ifeq ($(OS),Linux)
+  PORT ?= /dev/ttyACM0
+else ifeq ($(OS),Darwin)
+  PORT ?= $(shell ls -1 /dev/tty.SLAB_USBtoUART* | head -n 1)
```
Comment (thomaseichinger) [inline boards/nucleo-f334/Makefile.include]: Does this board require an external UART adapter?

Comment (LudwigOrtmann) [inline boards/nucleo-f334/Makefile.include]: I think so.

Comment (LudwigOrtmann) [inline boards/nucleo-f334/Makefile.include]: BTW: the UART pins are located on the little add-on board (labeled tx/rx) in the default configuration if I remember this correctly.

Comment (haukepetersen) [inline boards/nucleo-f334/Makefile.include]: Nope, it does not (at least on Linux Mint it works through the on-board st-link).

Comment (thomaseichinger) [inline boards/nucleo-f334/Makefile.include]: Then this is most probably the wrong port. I'd suspect it to be something like `/dev/tty.usbmodem*` but this needs to be verified.

Comment (LudwigOrtmann) [inline boards/nucleo-f334/Makefile.include]: OK, in my previous porting attempt it did ;)

**Comment by LudwigOrtmann** [comment]
Tested, `hello-world` and `default` work on my board.

**Comment by LudwigOrtmann** [comment]
@haukepetersen are you planning on adding any more features to this board?

**Comment by haukepetersen** [comment]
@LudwigOrtmann it is definitely planned, but not as part of this PR. I would suggest to add further feature in separate PRs. The plan is actually to get very well tested and complete support for this board, as this is the board-of-choice for the Ell-i products.

**Comment by haukepetersen** [comment]
@thomaseichinger do you have access to @LudwigOrtmann's board for testing the UART access?

**Comment by thomaseichinger** [comment]
@haukepetersen No I don't think so. @LudwigOrtmann?

**Comment by LudwigOrtmann** [comment]
@thomaseichinger I can bring the board to university on Friday or you could drop by my place and test it here ;)

**Comment by thomaseichinger** [comment]
@LudwigOrtmann Friday it is. ;)

**Comment by haukepetersen** [comment]
We also got two more boards borrowed from Ell-i, I will bring them tomorrow...

**Comment by thomaseichinger** [comment]
ok, verified, it is `/dev/tty.usbmodem*`

File: boards/nucleo-f334/dist/openocd.cfg [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# nucloe-f334 Target
+source [find interface/stlink-v2-1.cfg]
+
+transport select hla_swd
+
```
Comment (thomaseichinger) [inline boards/nucleo-f334/dist/openocd.cfg]: You can discard this file. It's the same content as in `board/st_nucleo_f334r8.cfg`

File: boards/nucleo-f334/dist/flash.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+echo "##"
+echo "## Flashing $1"
+echo "##"
+
+openocd -f "${RIOTBOARD}/${BOARD}/dist/openocd.cfg" \
```
Comment (thomaseichinger) [inline boards/nucleo-f334/dist/flash.sh]: change to `-f board/st_nucleo_f334r8.cfg`

File: boards/nucleo-f334/dist/debug-server.sh [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+
+echo "##"
+echo "## Starting debug server"
+echo "##"
+
+openocd -f "${RIOTBOARD}/${BOARD}/dist/openocd.cfg" \
```
Comment (thomaseichinger) [inline boards/nucleo-f334/dist/debug-server.sh]: change to `-f board/st_nucleo_f334r8.cfg`

**Comment by haukepetersen** [comment]
adapted to @thomaseichingers comments.

**Comment by LudwigOrtmann** [comment]
still works.

File: cpu/stm32f3/stm32f334r8_linkerscript.ld [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+OUTPUT_FORMAT("elf32-littlearm", "elf32-littlearm", "elf32-littlearm")
+OUTPUT_ARCH(arm)
+SEARCH_DIR(.)
+
+/* Memory Spaces Definitions */
+MEMORY
+{
+    rom (rx)    : ORIGIN = 0x08000000, LENGTH = 64K
+    ram (rwx)   : ORIGIN = 0x20000000, LENGTH = 12K
```
Comment (LudwigOrtmann) [inline cpu/stm32f3/stm32f334r8_linkerscript.ld]: The leaflet says 16K.

Comment (haukepetersen) [inline cpu/stm32f3/stm32f334r8_linkerscript.ld]: yapp, but the last 4K are buffered and I wasn't sure if they are mapped continuously onto the same address space. They are, so I will adjust this.

**Comment by haukepetersen** [comment]
adjusted linker script to utilize all 16K of RAM

File: boards/nucleo-f334/board.c [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    LED_GREEN_PORT->OSPEEDR |= 0x00000c00;
+    /* set output type to push-pull */
+    LED_GREEN_PORT->OTYPER &= ~(0x00000020);
+    /* configure pins as general outputs */
+    LED_GREEN_PORT->MODER &= ~(0x00000c00);
+    LED_GREEN_PORT->MODER |= 0x00000400;
+    /* disable pull resistors */
+    LED_GREEN_PORT->PUPDR &= ~(0x00000c00);
+    /* turn all LEDs off */
+    LED_GREEN_PORT->BRR = 0x00c0;
```
Comment (LudwigOrtmann) [inline boards/nucleo-f334/board.c]: Can we have constants for these numbers?

Comment (haukepetersen) [inline boards/nucleo-f334/board.c]: does not really make them better readable, with these numbers at least you can see on first sight which bits are set/cleared...

**Comment by haukepetersen** [comment]
so how about merging this?

**Comment by mehlis** [comment]
tested: uart and LED_GREEN is working

ACK -> merge

**Comment by mehlis** [comment]
@haukepetersen do the squash

**Comment by haukepetersen** [comment]
did.

**Comment by haukepetersen** [comment]
and go.

**Comment by OlegHahm** [comment]
> The story is: Hauke sitting at POINT kickoff meeting. Someone shows him the nucleo board, asks for RIOT support. Hauke says "no, but I bet I can make a port in 30min." After 27 minutes he had hello-world running on the board.

Awesome! (Dunno whether RIOT or Hauke...)

**Pull request merged**